### DISCUSSION TITLE
What is the intended way of starting the examples server?

### DISCUSSION THREAD

**Discussion by senritsu** [boardgameio/boardgame.io#984] [Q&A] [answered]
I have tried to run the remote multiplayer examples (chess, lobby) from `/examples/react-web`. Naturally they need a server instance running (on port `8000`).

While `server.js` exists, I could not figure out how the setup is meant to be used. 

# Attempts

## :x: Plain node

Simply running the server via Node does not work because the dependencies use ES imports. I assume it could work when using `type: 'module'` in `package.json` and maybe some additional fiddling, but I did not examine this further.

## 🤔 Similar to docs

Using `esm` as per [the example in the docs](https://boardgame.io/documentation/#/multiplayer?id=setting-up-the-server) works, but needs additional setup

- `npm install esm`
- new script in `package.json` for convenience
- apparently a change in `server.js`:

```js
// status quo in the repo, does not work because esm does not resolve `alias` in package.json
import { Server, Origins } from 'boardgame.io/server';
// works, but relative imports are kind of "eh"
import { Server, Origins } from '../../dist/cjs/server';
```

`npm i boardgame.io` and removing the `alias` from `package.json` would probably also work, but removes the option of using local changes to the library, which is not ideal. Using other options of aliasing modules like [module-alias](https://www.npmjs.com/package/module-alias) or similar might work, but I did not go there.

## ❓ Parcel

I tried running the server using Parcel, in the hopes of side-stepping the alias issue, but while Parcel does not complain, the server also does not work. Not quite sure if Parcel is fundamentally not suitable for server-side code, or if I am doing something wrong. I don't have noteworthy prior experience with Parcel apart from running a sample or two some years ago (more of a webpack person, not necessarily by choice). A quick google did not provide much relevant information, although it seems (sort of obviously) primarily geared towards web app bundling with html entry points.

Additionally, `dist/esm` does not include the server code (which is only built as CJS), so the `alias` in `package.json` does not work and the imports need to be changed to `../../cjs/server` like when using `esm`.

# Conclusion

Either I am missing something obvious, or this could use a `README.md` (or both). I would be happy to write one if someone gives me some pointers towards the intended workflow.

Apart from that it was really comfortable to get everything running in VS Code using Dev containers. Has it been considered to bundle a preconfigured `devcontainer.json` to define a canonical dev environment for easy contribution?

**Comment by delucis** [comment, marked as answer]
Hey! Here are the instructions [from the main README](https://github.com/boardgameio/boardgame.io#running-examples-in-this-repository):

> ### Running examples in this repository
>
> ```sh
> npm install
> npm start
> ```

That both starts the front-end server and the game server as parallel processes. For reference, this is the actual command used to run just the server, but it’s less friendly looking 😆 :
https://github.com/boardgameio/boardgame.io/blob/c7c3fab9022c344ac353e72ad0d3222b9b9bbb8c/package.json#L12-L12

> Apart from that it was really comfortable to get everything running in VS Code using Dev containers. Has it been considered to bundle a preconfigured devcontainer.json to define a canonical dev environment for easy contribution?

I’m not familiar with dev containers. Happy to take a look at a PR adding that JSON config if it would be helpful. Do you have a good link with more information about that feature?

**Reply by senritsu** [reply]
Well ok then, it seems I was so distracted by the examples sources and especially the `package.json` in the examples folder, I didn't even look properly in the main package file or README 🤦‍♂️

> Do you have a good link with more information about that feature?

[here](https://code.visualstudio.com/docs/remote/containers#_quick-start-open-a-git-repository-or-github-pr-in-an-isolated-container-volume) is the quick start from Microsoft I used. I used it for the first time myself, but it was fairly trivial to get working. Being on a fresh windows install without Node and everything, I figured this is the ideal time to try it out.

With some minimal configuration, it looks quite possible to start contributing with almost-zero setup on the local PC (only installing Docker, VS Code and git), which is kind of nice. It also avoids all the idiosyncrasies that pop up when developing Node projects on Windows, which is also kind of nice.

**Reply by senritsu** [reply]
It also seems I cannot mark your comment as an answer, I have a stuck popup explaining how marking answers works (which gets a 422 response on clicking "Got It!"), and the button "Mark as answer" does nothing 🤔

**Reply by delucis** [reply]
> It also seems I cannot mark your comment as an answer, I have a stuck popup explaining how marking answers works (which gets a 422 response on clicking "Got It!"), and the button "Mark as answer" does nothing 🤔

Odd! Guess that’s a GitHub bug. I’ve marked it as the answer.

**Reply by delucis** [reply]
Thanks for the link. I’ll take a look.

**Reply by senritsu** [reply]
:+1: Curious how the next time I use Q&A will work out :)
